### PR TITLE
Function Resource Auth

### PR DESCRIPTION
**Pull request by tsurdilo**
Signed-off-by: Tihomir Surdilovic <tihomir@temporal.io>

**Many thanks for submitting your Pull Request :heart:!**

**Please specify parts this PR updates:**

- [x] Specification
- [x] Schema
- [x] Examples
- [ ] Extensions
- [ ] Roadmap
- [ ] Use Cases
- [ ] Community
- [ ] TCK
- [ ] Other

**What this PR does / why we need it**:
* Adds the "auth" property to top-level workflow definition. This prop includes reusable auth info for resources that are accessed by operation prop of function definitions. It does not provide auth info for the invocation of the function, but only info how to access the definition file that includes this information. 
* Adds the "authRef" property to the function definition. This property is used to reference the auth definition by its name
* Adds the missing metadata prop to function definitions 
**Special notes for reviewers**:
This is still a Work In progress
**Additional information (if needed):**

### CODE REVIEW

**Comment by tsurdilo** [comment]
@cdavernas 
This is the initial PR ..we need to work on defining the properties for each auth type.
Please help with that :)

File: schema/auth.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          "description": "Unique auth name",
+          "minLength": 1
+        },
+        "type": {
+          "type": "string",
+          "description": "Defines the auth type",
+          "enum": [
+            "basic",
+            "bearer",
+            "openid"
```
Comment (cdavernas) [inline schema/auth.json:43]: We should probably change that to `oidc`

File: schema/auth.json [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  "definitions": {
+    "authdef": {
+      "type": "object",
+      "properties": {
+        "name": {
+          "type": "string",
+          "description": "Unique auth name",
+          "minLength": 1
+        },
+        "type": {
```
Comment (cdavernas) [inline schema/auth.json:37]: We should probably rename that to `scheme`, which is, IMHO, more ubiquitous

File: schema/auth.json [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        "additionalItems": false,
+        "minItems": 1
+      }
+    ]
+  },
+  "required": [
+    "auth"
+  ],
+  "definitions": {
+    "authdef": {
```
Comment (cdavernas) [inline schema/auth.json:29]: I think we need an `anyOf` here, with the following schemas:

## Basic
```
{
  "name": "MyBasicAuth",
  "scheme": "basic",
  "properties": {
    "username": "MyUserName",
    "password":  "MyPassword"
  }
}
```

## Bearer

```
{
  "name": "MyBearerAuth",
  "scheme": "bearer",
  "properties": {
    "token": "MyToken"
  }
}
```

## OpenID Connect

```
{
  "name": "MyBearerAuth",
  "scheme": "oidc",
  "properties": {
    "grantType": "resourceOwner | clientCredentials | tokenExchange",
    "clientId": "MyClientId", #required
    "clientSecret": "MyClientSecret", #optional
    "username": "MyUserName", #only for resourceOwner
    "password": "MyPassword" #only for resourceOwner,
    "scope": "openid otherScopes"
  }
}
```

**Comment by tsurdilo** [comment]
@cdavernas update the schema. if its ok ill finish the specification doc next. let me know

File: schema/auth.json (hunk 2 above)
Comment (tsurdilo) [reply schema/auth.json]: done

File: schema/auth.json (hunk 3 above)
Comment (tsurdilo) [reply schema/auth.json:29]: added

File: schema/auth.json (hunk 1 above)
Comment (tsurdilo) [reply schema/auth.json]: done

File: schema/auth.json (hunk 3 above)
Comment (tsurdilo) [reply schema/auth.json:29]: only difference is that instead of "properties" i called it "info"

File: schema/auth.json [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+          "type": "string",
+          "description": "Defines the auth type",
+          "enum": [
+            "basic",
+            "bearer",
+            "oidc"
+          ],
+          "default": "basic"
+        },
+        "info": {
```
Comment (cdavernas) [inline schema/auth.json:47]: In the end, I'm really not a big fan of the term `info`. IMHO, and according to some of my searches, `properties` is the right term to use in this case, or at least it's the nomenclature used on .NET Framework. WDYT?

File: schema/auth.json [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+              "type": "string",
+              "description": "String or a workflow expression. Contains the client id",
+              "minLength": 1
+            },
+            "clientSecret": {
+              "type": "string",
+              "description": "String or a workflow expression. Contains the client secret",
+              "minLength": 1
+            },
+            "scope": {
```
Comment (cdavernas) [inline schema/auth.json:155]: As OIDC allows settings multiple space-separated scopes, maybe we should rename the property `scopes` and turn it into an array of strings. WDYT?

File: schema/auth.json [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    "oidcinfodef": {
+      "oneOf": [
+        {
+          "type": "string",
+          "description": "Expression referencing a workflow secret that contains all needed oidc auth info"
+        },
+        {
+          "type": "object",
+          "description": "Oidc auth information",
+          "properties": {
```
Comment (cdavernas) [inline schema/auth.json:135]: I'm sorry I just realized while implementing it that I forgot some properties (mainly used by the [token-exchange](https://datatracker.ietf.org/doc/html/rfc8693) grant type):
- **audiences**: array of string
- **subjectToken**: string
- **requestedSubject**: string
- **requestedIssuer**: string

**Review by cdavernas — COMMENTED** [review]
Thank you sooooooooooo much for this PR @tsurdilo!!!
Approved IMHO, aside from a couple of comments ;)

**Comment by cdavernas** [comment]
Ahhh! Final thought: could we add metadata to all `authinfodef`? Might be usefull for some exotic use cases

File: schema/auth.json (hunk 4 above)
Comment (tsurdilo) [reply schema/auth.json]: updated

File: schema/auth.json (hunk 5 above)
Comment (tsurdilo) [reply schema/auth.json]: updated

File: schema/auth.json (hunk 6 above)
Comment (tsurdilo) [reply schema/auth.json:141]: added

**Comment by tsurdilo** [comment]
> Ahhh! Final thought: could we add metadata to all `authinfodef`? Might be usefull for some exotic use cases

added

**Comment by tsurdilo** [comment]
rebased

**Comment by cdavernas** [comment]
Arf... Sorry, I just realized when trying to run my tests that we still miss an `Authority` uri property for the `oidcauthenticationproperties`

**Comment by tsurdilo** [comment]
can u show me example?

File: schema/auth.json [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
```
Comment (cdavernas) [inline schema/auth.json:141]: `authority: https://keycloak.local`
The runtime then knows to contact https://keycloak.local/.well-known/openid-configuration to get the file that describe the openid endpoints

File: schema/auth.json [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+          "description": "Unique auth name",
+          "minLength": 1
+        },
+        "scheme": {
+          "type": "string",
+          "description": "Defines the auth type",
+          "enum": [
+            "basic",
+            "bearer",
+            "oidc"
```
Comment (cdavernas) [inline schema/auth.json:43]: You will hate me, but `OIDC` is not the right term, in the end. That became obvious to me when reading the [token exchange RFC](https://datatracker.ietf.org/doc/html/rfc8693), where it is clearly stated that token exchange is an `oauth2` grant type.
So yeah, we should replace `oidc` by `oauth2`.
So sorry about that, it's already the second time I'm asking to rename those 😒

File: schema/auth.json [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        },
+        {
+          "type": "object",
+          "description": "Oidc auth information",
+          "properties": {
+            "grantType": {
+              "type": "string",
+              "description": "Defines the auth type",
+              "enum": [
+                "resourceOwner",
```
Comment (cdavernas) [inline schema/auth.json:146]: Even though I don't like naming it that way, I think we probably should change that to `password`, which is the actual value you have to input when requesting a token with the **Resource Owner Password Credentials** grant type.

File: specification.md [hunk 10]
```diff
@@ -1925,6 +1955,61 @@ Note that using metadata for cases such as above heavily reduces the portability
...
+it defines an URI to a resource containing auth definitions (json or yaml).
+If `object` type, it has the following properties:
+
+| Parameter | Description | Type | Required |
+| --- | --- | --- | --- |
+| name | Unique function name | string | yes |
+| type | Auth type | enum | yes |
+| [metadata](#Workflow-Metadata) | Metadata information | object | no |
+
+TODO - Finish definition parameters !
```
Comment (ricardozanini) [inline specification.md:1970]: Is this a leftover?

**Review by ricardozanini — APPROVED** [review]

File: specification.md (hunk 10 above)
Comment (tsurdilo) [reply specification.md:1970]: yeh - nice catch . will remove

**Comment by tsurdilo** [comment]
rebased

File: schema/auth.json [hunk 11]
```diff
@@ -0,0 +1,214 @@
+{
+  "$id": "https://serverlessworkflow.io/schemas/0.7/auth.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "description": "Serverless Workflow specification - auth schema",
+  "type": "object",
+  "auth": {
```
[the remark this answers is not in the document]
Comment (tsurdilo) [reply schema/auth.json:6]: hmm not sure to me this is weird 
cause dont think you say "workflow authentications information" but singular "workflow authentication information"

[the remark this answers is not in the document]
Comment (cdavernas) [reply schema/auth.json:6]: Yea but IMHO, it's rather action authentications, isn't it?

File: schema/auth.json (hunk 9 above)
Comment (tsurdilo) [reply schema/auth.json:146]: hmm i am not expert in this but this enum values look confusing to me :) 

             "enum": [
                "password",
                "clientCredentials",
                "tokenExchange"
              ],

is this ok ?

File: schema/auth.json (hunk 7 above)
Comment (tsurdilo) [reply schema/auth.json:141]: added

File: schema/auth.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (tsurdilo) [reply schema/auth.json:6]: i havent seen plural used for "authentication" ..but its not my native language so im probably wrong :) 
for me its still "action authentication" :) if its one or many

File: schema/auth.json (hunk 8 above)
Comment (tsurdilo) [reply schema/auth.json]: updated

File: schema/auth.json (hunk 9 above)
Comment (tsurdilo) [reply schema/auth.json]: updated

**Comment by tsurdilo** [comment]
@cdavernas added the suggestions, except the "auth/s" thing which i am honestly not sure about but can update no problem :) 

Can you check the auth schema now to see if the props are ok and then i can work on finishing the specification text for this

File: schema/auth.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (cdavernas) [reply schema/auth.json:6]: @tsurdilo I'm a native french speaker, can't argue with you here 🤣 
It was just something I had seen out there, but again they might not be native english speakers either...

File: schema/auth.json (hunk 9 above)
Comment (cdavernas) [reply schema/auth.json]: @tsurdilo Yes, thanks!!!

**Review by cdavernas — APPROVED** [review]
@tsurdilo LGTM!!! Many thanks!!!

**Comment by tsurdilo** [comment]
@cdavernas thanks! will finish the spec section and ping you again for review of that.
Yeah we need to find someone that speaks the queens english to help us out with the auth/auths thing :)

**Pull request merged**